(Mac) error: failed to run custom build command for `pgx-pg-sys v0.1.21`

I following this [Getting Started](https://github.com/zombodb/pgx#getting-started) guide, when running command `cargo pgx run pg13` got this errors:

(macOS 11.3.1, Xcode 12.5, rustc 1.54.0-nightly)

```
error: failed to run custom build command for `pgx-pg-sys v0.1.21`

Caused by:
  process didn't exit successfully: `/Users/hello/work/lab/myfile_fdw/target/debug/build/pgx-pg-sys-7e5dcc9e92761c9a/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=/Users/hello/.pgx/config.toml
  cargo:rerun-if-changed=include/pg10.h
  cargo:rerun-if-changed=include/pg11.h
  cargo:rerun-if-changed=include/pg12.h
  cargo:rerun-if-changed=include/pg13.h
  cargo:rerun-if-changed=cshim/pgx-cshim.c
  cargo:rerun-if-changed=cshim/Makefile

  --- stderr
  manifest_dir=/Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21
  shim_dir=/Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21/cshim
  Generating bindings for pg13
  Generating bindings for pg10
  Generating bindings for pg11
  Generating bindings for pg12
  /Users/hello/.pgx/13.3/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found
  /Users/hello/.pgx/13.3/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found, err: true
  thread '<unnamed>' panicked at 'Unable to generate bindings for pg13: ()', /Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21/build.rs:486:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  /Users/hello/.pgx/11.12/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found
  /Users/hello/.pgx/11.12/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found, err: true
  thread '<unnamed>' panicked at 'Unable to generate bindings for pg11: ()', /Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21/build.rs:486:13
  /Users/hello/.pgx/12.7/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found
  /Users/hello/.pgx/12.7/pgx-install/include/postgresql/server/c.h:59:10: fatal error: 'stdio.h' file not found, err: true
  thread '<unnamed>' panicked at 'Unable to generate bindings for pg12: ()', /Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21/build.rs:486:13
  /Users/hello/.pgx/10.17/pgx-install/include/postgresql/server/c.h:81:10: fatal error: 'stdio.h' file not found
  /Users/hello/.pgx/10.17/pgx-install/include/postgresql/server/c.h:81:10: fatal error: 'stdio.h' file not found, err: true
  thread '<unnamed>' panicked at 'Unable to generate bindings for pg10: ()', /Users/hello/.cargo/registry/src/github.com-1ecc6299db9ec823/pgx-pg-sys-0.1.21/build.rs:486:13
      [error] failed to build extension
```

The command `cargo pgx init` was successful:

```
  Discovered Postgres v13.3, v12.7, v11.12, v10.17
  Downloading Postgres v10.17 from https://ftp.postgresql.org/pub/source/v10.17/postgresql-10.17.tar.bz2
  Downloading Postgres v11.12 from https://ftp.postgresql.org/pub/source/v11.12/postgresql-11.12.tar.bz2
  Downloading Postgres v12.7 from https://ftp.postgresql.org/pub/source/v12.7/postgresql-12.7.tar.bz2
  Downloading Postgres v13.3 from https://ftp.postgresql.org/pub/source/v13.3/postgresql-13.3.tar.bz2
     Removing /Users/hello/.pgx/13.3
     Removing /Users/hello/.pgx/12.7
    Untarring Postgres v13.3 to /Users/hello/.pgx/13.3
    Untarring Postgres v12.7 to /Users/hello/.pgx/12.7
  Configuring Postgres v12.7
  Configuring Postgres v13.3
     Removing /Users/hello/.pgx/11.12
    Untarring Postgres v11.12 to /Users/hello/.pgx/11.12
  Configuring Postgres v11.12
     Removing /Users/hello/.pgx/10.17
    Untarring Postgres v10.17 to /Users/hello/.pgx/10.17
  Configuring Postgres v10.17
    Compiling Postgres v13.3
    Compiling Postgres v11.12
    Compiling Postgres v12.7
    Compiling Postgres v10.17
   Installing Postgres v11.12 to /Users/hello/.pgx/11.12/pgx-install
   Installing Postgres v12.7 to /Users/hello/.pgx/12.7/pgx-install
   Installing Postgres v13.3 to /Users/hello/.pgx/13.3/pgx-install
   Installing Postgres v10.17 to /Users/hello/.pgx/10.17/pgx-install
   Validating /Users/hello/.pgx/10.17/pgx-install/bin/pg_config
   Validating /Users/hello/.pgx/11.12/pgx-install/bin/pg_config
   Validating /Users/hello/.pgx/12.7/pgx-install/bin/pg_config
   Validating /Users/hello/.pgx/13.3/pgx-install/bin/pg_config
The last command took 418.053 seconds.
```

--- a/pgx-pg-config/src/lib.rs
+++ b/pgx-pg-config/src/lib.rs
@@ -14,6 +14,7 @@ use std::fmt::{self, Display, Formatter};
 use std::process::Stdio;
 use std::{
     collections::HashMap,
+    ffi::OsString,
     io::ErrorKind,
     path::{Path, PathBuf},
     process::Command,
@@ -246,6 +247,10 @@ impl PgConfig {
 
     pub fn sharedir(&self) -> eyre::Result<PathBuf> {
         Ok(self.run("--sharedir")?.into())
+    }
+
+    pub fn cppflags(&self) -> eyre::Result<OsString> {
+        Ok(self.run("--cppflags")?.into())
     }
 
     pub fn extension_dir(&self) -> eyre::Result<PathBuf> {

--- a/pgx-pg-sys/Cargo.toml
+++ b/pgx-pg-sys/Cargo.toml
@@ -50,3 +50,4 @@ syn = { version = "1.0.101", features = [ "extra-traits", "full", "fold", "parsi
 eyre = "0.6.8"
 color-eyre = "0.6.2"
 rustversion = "1.0"
+shlex = "1.1.0"

--- a/pgx-pg-sys/Cargo.toml
+++ b/pgx-pg-sys/Cargo.toml
@@ -50,4 +50,4 @@ syn = { version = "1.0.101", features = [ "extra-traits", "full", "fold", "parsi
 eyre = "0.6.8"
 color-eyre = "0.6.2"
 rustversion = "1.0"
-shlex = "1.1.0"
+shlex = "1.1.0" # shell lexing, also used by many of our deps

--- a/pgx-pg-sys/build.rs
+++ b/pgx-pg-sys/build.rs
@@ -679,7 +679,7 @@ fn extra_bindgen_clang_args(pg_config: &PgConfig) -> eyre::Result<Vec<String>> {
                     // The logic we'd like ideally is for `cargo pgx init` to
                     // choose a good SDK in the first place, and force postgres
                     // to use it. Then, the logic in this build script would
-                    // Just Work without changes (since we are using it's
+                    // Just Work without changes (since we are using its
                     // sysroot verbatim).
                     //
                     // The value of "Good" here is tricky, but the logic should
@@ -707,7 +707,7 @@ fn extra_bindgen_clang_args(pg_config: &PgConfig) -> eyre::Result<Vec<String>> {
                     // SDK version, but it's not an option for postgres.
                     let major_version = pg_config.major_version()?;
                     println!(
-                        "cargo:warning=postgres version {major_version} was compiled against a \
+                        "cargo:warning=postgres v{major_version} was compiled against an \
                          SDK Root which does not seem to exist on this machine ({}). You may \
                          need to re-run `cargo pgx init` and/or update your command line tools.",
                         pair[1],

--- a/pgx-pg-sys/build.rs
+++ b/pgx-pg-sys/build.rs
@@ -535,6 +535,7 @@ fn run_bindgen(pg_config: &PgConfig, include_h: &PathBuf) -> eyre::Result<syn::F
     let bindings = bindgen::Builder::default()
         .header(include_h.display().to_string())
         .clang_arg(&format!("-I{}", includedir_server.display()))
+        .clang_args(&extra_bindgen_clang_args(pg_config)?)
         .parse_callbacks(Box::new(PgxOverrides::default()))
         .blocklist_type("Datum") // manually wrapping datum types for correctness
         .blocklist_type("NullableDatum")
@@ -632,6 +633,92 @@ fn build_shim_for_version(
     }
 
     Ok(())
+}
+
+fn extra_bindgen_clang_args(pg_config: &PgConfig) -> eyre::Result<Vec<String>> {
+    let mut out = vec![];
+    if std::env::var("CARGO_CFG_TARGET_OS").unwrap_or_default() == "macos" {
+        // On macOS, find the `-isysroot` arg out of the c preprocessor flags,
+        // to handle the case where bindgen uses a libclang isn't provided by
+        // the system.
+        let flags = pg_config.cppflags()?;
+        // In practice this will always be valid UTF-8 because of how the
+        // `pgx-pg-config` crate is implemented, but even if it were not, the
+        // problem won't be with flags we are interested in.
+        let flags = shlex::split(&flags.to_string_lossy()).unwrap_or_default();
+        // Find the `-isysroot` flags -- The rest are `-I` flags that don't seem
+        // to be needed inside the code (and feel likely to cause bindgen to
+        // emit bindings for unrelated libraries)
+        for pair in flags.windows(2) {
+            if pair[0] == "-isysroot" {
+                if std::path::Path::new(&pair[1]).exists() {
+                    out.extend(pair.into_iter().cloned());
+                } else {
+                    // The SDK path doesnt exist. Emit a warning, which they'll
+                    // see if the build ends up failing (it may not fail in all
+                    // cases, so we don't panic here).
+                    //
+                    // There's a bunch of smarter things we can try here, but
+                    // most of them either break things that currently work, or
+                    // are very difficult to get right. If you try to fix this,
+                    // be sure to consider cases like:
+                    //
+                    // - User may have CommandLineTools and not Xcode, vice
+                    //   versa, or both installed.
+                    // - User may using a newer SDK than their OS, or vice
+                    //   versa.
+                    // - User may be using a newer SDK than their XCode (updated
+                    //   Command line tools, not OS), or vice versa.
+                    // - And so on.
+                    //
+                    // These are all actually fairly common. Note that the code
+                    // as-is is *not* broken in these cases (except on OS/SDK
+                    // updates), so care should be taken to avoid changing that
+                    // if possible.
+                    //
+                    // The logic we'd like ideally is for `cargo pgx init` to
+                    // choose a good SDK in the first place, and force postgres
+                    // to use it. Then, the logic in this build script would
+                    // Just Work without changes (since we are using it's
+                    // sysroot verbatim).
+                    //
+                    // The value of "Good" here is tricky, but the logic should
+                    // probably:
+                    //
+                    // - prefer SDKs from the CLI tools to ones from XCode
+                    //   (since they're guaranteed compatible with the user's OS
+                    //   version)
+                    //
+                    // - prefer SDKs that specify only the major SDK version
+                    //   (e.g. MacOSX12.sdk and not MacOSX12.4.sdk or
+                    //   MacOSX.sdk), to avoid breaking too frequently (if we
+                    //   have a minor version) or being totally unable to detect
+                    //   what version of the SDK was used to build postgres (if
+                    //   we have neither).
+                    //
+                    // - Avoid choosing an SDK newer than the user's OS version,
+                    //   since postgres fails to detect that they are missing if
+                    //   you do.
+                    //
+                    // This is surprisingly hard to implement, as the
+                    // information is scattered across a dozen ini files.
+                    // Presumably Apple assumes you'll use
+                    // `MACOSX_DEPLOYMENT_TARGET`, rather than basing it off the
+                    // SDK version, but it's not an option for postgres.
+                    let major_version = pg_config.major_version()?;
+                    println!(
+                        "cargo:warning=postgres version {major_version} was compiled against a \
+                         SDK Root which does not seem to exist on this machine ({}). You may \
+                         need to re-run `cargo pgx init` and/or update your command line tools.",
+                        pair[1],
+                    );
+                };
+                // Either way, we stop here.
+                break;
+            }
+        }
+    }
+    Ok(out)
 }
 
 fn run_command(mut command: &mut Command, version: &str) -> eyre::Result<Output> {

